jobs/gtfs-rt-parser/gtfs_rt_parser.py

```diff
@@ -123,109 +218,440 @@ def identify_files(glob, rt_file_type: RTFileType, progress=False) -> List[RTFil
+    fs,
+    dst_path_rt,
+    tmp_dir,
+    hour: RTAggregation,
+    verbose=False,
+    pbar=None,
+):
+    written = 0
+    errors = []
+    gzip_fname = str(tmp_dir + f"/data_{hour.suffix}" + JSONL_GZIP_EXTENSION)
+    with gzip.open(gzip_fname, "w") as gzipfile:
+        for rt_file in hour.source_files:
+            feed = gtfs_realtime_pb2.FeedMessage()
+
+            try:
+                with open(
+                    os.path.join(dst_path_rt, rt_file.timestamped_filename), "rb"
+                ) as f:
+                    feed.ParseFromString(f.read())
+                parsed = json_format.MessageToDict(feed)
+            except DecodeError as e:
+                errors.append(
+                    RTFileProcessingError(
+                        step="parse",
+                        exception=str(e),
+                        file=rt_file,
+                    )
+                )
+                continue
+
+            if not parsed or "entity" not in parsed:
+                if verbose:
+                    log(
+                        f"WARNING: no records found in {str(rt_file.path)}",
+                        fg=typer.colors.YELLOW,
+                        pbar=pbar,
+                    )
+                continue
+
             for record in parsed["entity"]:
-                record.update(
-                    {
-                        "header": parsed["header"],
-                        "metadata": json.loads(
-                            rt_file.json()
-                        ),  # back and forth so we use pydantic serialization
-                    }
+                gzipfile.write(
+                    (
+                        json.dumps(
+                            {
+                                "header": parsed["header"],
+                                # back and forth so we use pydantic serialization
+                                "metadata": json.loads(rt_file.json()),
+                                **copy.deepcopy(record),
+                            }
+                        )
+                        + "\n"
+                    ).encode("utf-8")
                 )
-                gzipfile.write((json.dumps(record) + "\n").encode("utf-8"))
                 written += 1
+            del parsed
 
-        out_path = f"{rt_file.hive_path(bucket)}{EXTENSION}"  # probably a better way to do this
-        msg = f"writing {written} lines from {str(rt_file.path)} to {out_path}"
-        if pbar:
-            pbar.write(msg)
-        else:
-            typer.echo(msg)
-        put_with_retry(fs, gzip_fname, out_path)
-        if pbar:
-            pbar.update(1)
+    if written:
+        log(
+            f"writing {written} lines to {hour.data_hive_path}{JSONL_GZIP_EXTENSION}",
+            pbar=pbar,
+        )
+        put_with_retry(fs, gzip_fname, f"{hour.data_hive_path}{JSONL_GZIP_EXTENSION}")
+    else:
+        log(
+            f"WARNING: no records at all for {hour.data_hive_path}",
```

Comment:
**question**: This kind of thing is making me wonder.... Is there any value in writing out a more generic metadata file that lists the feeds that were validated or something, with the outcome? As you've mentioned to me, no files found can be because there were no errors or because validation didn't run. Is there a way for us to distinguish between those cases / do we need to be able to do that?

Reply:
I guess I'm wondering if there's value in refactoring the `error` files to just be `outcome` files so that you have like a record of what feeds were operated on. Can a single feed throw more than one error (within the context of just parsing or just validating? Like I assume a feed can throw an error on parsing AND on validation, but within parsing is it possible to have multiple errors written? If not, think we could just log an error outcome with whatever the error was.)

Basically... Refactor the `RTError` class to be like an `RTOutcome` class that gets created either way?